### ISSUE TITLE
CUDA optimization of THCTensorMathPointwise.cuh

### ISSUE THREAD

**Issue by ThisIsIsaac** [pytorch/pytorch#22612]
@zou3519 

pytorch's pointwise operations inside [THCTensorMathPointwise.cuh](https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorMathPointwise.cuh) seems like it could use some optimizations for `int8` and `half` types. For types that are smaller than 4 bytes, there are significant amounts of bandwidth wasted. By having each thread load 4 bytes -- i.e. 4 `int8` elements or 2 `half` elements -- and process them at once, we can eliminate wasted bandwidth.

**Comment by fmassa** [comment]
Note that `THCTensorMathPointwise` is kind of deprecated in favour of TensorIterator, which for CUDA the loops can be found in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Loops.cuh

**Comment by colesbury** [comment]
The same issue occurs in the CUDA implementations of TensorIterator, but I don't think it's worth the code and implementation complexity to optimize for these small data types.

Note that we still get good performance on half and short integers due to caching of global loads (L1 in Pascal + Volta, L2 in older architectures):

bandwidth for addition on 40 MiB on P100:
float32: 465 GiB/sec
float16: 421 GiB/sec
int16: 428 GiB/sec
int8: 285 GiB/sec

https://gist.github.com/colesbury/41fc51225f25e72d7dd0049ec3966e04

I haven't seen any applications where elementwise operations on byte-sized data is a bottleneck. Generally bytes (or bools) are used in combination with some other data type like in masking or indexing.

**Comment by ThisIsIsaac** [comment]
Thanks for the detailed review, @colesbury. 

My colleage -- @minty99 -- profiled our DNN program and found that `kernelPointwiseApply2::CopyOp` is the bottleneck. The DNN program we are optimizing is Nvidia's Waveglow -- state of the art text-to-speech DNN algorithm. Here is a screenshot from Nsight Systems of Waveglow. I have also attached the profiling output file as well.
[waveglow_long_benchmark.qdrep.zip](https://github.com/pytorch/pytorch/files/3375531/waveglow_long_benchmark.qdrep.zip)


<img width="968" alt="Screen Shot 2019-07-10 at 11 38 49 AM" src="https://user-images.githubusercontent.com/49523965/60936231-594fa080-a307-11e9-88e0-e6034626705d.png">

We can see that we spend about 35% of time on `CopyOp`.

Our plan is to optimize the `CopyOp` kernel for `fp16` types by loading 2 elements, instead of 1, at a time. We would have to condition on the type and the size of the matrix -- the type has to be `half` and the number of elements has to be a multiple of 2. Otherwise, we should call the original unoptimized kernel.

**Comment by ThisIsIsaac** [comment]
Is `CopyOp` deprecated? It is used in Pytorch 1.0, but I'm not sure if it is still used.

**Comment by fmassa** [comment]
@ThisIsIssac FYI I'd recommend you profiling with latest master version of PyTorch. The copy operation on the GPU has been recently moved to use TensorIterator https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Copy.cu

**Comment by ThisIsIsaac** [comment]
@fmassa Thanks for the clarification. The problem is Waveglow requires Pytorch apex plugin, but it doesn't work with the newest pytorch. 

So if we were to optimize `CopyOp` in [THCTensorCopy.cu](https://github.com/pytorch/pytorch/blob/3a8d7463bd024dcbd00d04a323cdb10dfe867f18/aten/src/THC/THCTensorCopy.cu), then it wouldn't be of any help to Pytorch. Am I correct?

**Comment by fmassa** [comment]
Exactly, this code path is not used anymore

**Comment by minty99** [comment]
How can I profile `kernelPointwiseApply2::CopyOp` only? @colesbury already uploaded great example for addition, but I can't find appropriate method to test CopyOp.

**Comment by fmassa** [comment]
@minty99 I believe you might need to switch back to an old version of PyTorch, as I believe we have completely removed `CopyOp` in https://github.com/pytorch/pytorch/pull/20685

**Comment by minty99** [comment]
I'm searching for test code using Pytorch 1.0.0's `CopyOp` in Python, like `x1 + x2` was used to test addition.

**Comment by fmassa** [comment]
@minty99 try `x1.copy_(x2)`

**Comment by minty99** [comment]
It uses `kernelPointwiseApply1` with `TensorFillOp`. Our bottleneck was `kernelPointwiseApply2` with `CopyOp`.

**Comment by fmassa** [comment]
Are you sure that the following code
```python
a = torch.rand(2)
b = torch.rand(2)
a.copy_(b)
```
dispatches to `TensorFillOp`?

**Comment by minty99** [comment]
Previously I used torch.zeros. It was dispatched to `TensorFillOp`.

Otherwise, Your code dispatches to `generate_uniform`.
I tested by this code:
```
import torch

for i in range(10):
        a = torch.rand(10000, device='cuda', dtype=torch.float16)
        b = torch.rand(10000, device='cuda', dtype=torch.float16)
        a.copy_(b)
        torch.cuda.synchronize()
```

**Comment by fmassa** [comment]
@minty99 you are just checking the wrong call. `torch.rand` dispatches to `generate_uniform`, but `a.copy_(b)` should dispatch to the copy kernel

**Comment by ezyang** [comment]
Marking this as low priority as we may have possibly already fixed this on master. I'm not closing until we confirm though.

**Comment by colesbury** [comment]
CopyOp (and its replacement in the master branch [1]) are only executed for cases that can't use `cudaMemcpyAsync`, such as non-contiguous copies or copies that cast between different data types. For example @minty99's example of copying 10000 elements to 10000 will use `cudaMemcpyAsync` instead of CopyOp. `cudaMemcpy` calls show up differently in nvprof IIRC.

Something like the following will use CopyOp:

```
import torch
x = torch.empty(1000, 1000, device='cuda')
y = torch.empty(1000, device='cuda')
x.copy_(y)
```

Copies that convert between data types also will use CopyOp, but show up as something like CopyOp<c10::Half, float>. In your profile, these copy conversions look like they take <1%.

[1] The replacement has a terribly long and mangled C++ name that begins with "elementwise_kernel"

**Comment by minty99** [comment]
I checked @colesbury's code and it works great! Thanks for your help.
I got this result for CopyOp - however it was deprecated -
float32 164.61 GB/s
float16 84.39 GB/s
int16 82.99 GB/s
int8 42.37 GB/s

**Comment by ThisIsIsaac** [comment]
@ezyang , @fmassa  Okay, so we've caught up with the head of master branch's implementation of broadcasting copying. From profiling `float16` and `float32`, we still see some rooms for improvement. `float16` displays ** throughput 76% of `float32`'s throughput** which means we **could speed up broadcasting `float16` by ~30%**.

For waveglow which spends at least ~20% of time in broadcasting and element-wise addition of `fp16`, we will give this a shot.

Below is the IPython Notebook code we used to profile:

```python

# Initialization
torch.set_num_threads(1)
torch.randn(1, device='cuda:0')


def loop(f):
  for _ in range(30):
    f()
  torch.cuda.synchronize()

# time (s) = 0.0947s
# gld (GB) = 0.000001024 GB
# gst (GB) = 2GB
# throughput (GB/s) = 2GB * 30 / 0.0947s = 633GB/s
print('CUDA broadcasting 2GB [fp16]')
dst_fp16 = torch.randn(1024, 1024, 1024, device='cuda', dtype=torch.float16)
src_fp16 = torch.randn(1024, device='cuda', dtype=torch.float16)
%timeit loop(lambda: dst_fp16.copy_(src_fp16))

# time (s) = 0.145s
# gld (GB) = 0.000002048 GB
# gst (GB) = 4 GB
# throughput (GB/s) = 4GB * 30 / 0.145s = 827GB/s
print('CUDA broadcasting 4GB [fp32]')
dst_fp32 = torch.randn(1024, 1024, 1024, device='cuda', dtype=torch.float32)
src_fp32 = torch.randn(1024, device='cuda', dtype=torch.float32)
%timeit loop(lambda: dst_fp32.copy_(src_fp32))
``` 

We will likely have to case on template types and the size of the tensor (the size must be divisible by 2). 

I am going to open another issue as we are optimizing a different kernel.

**Comment by fmassa** [comment]
@ThisIsIsaac on which GPUs are you testing this?

On a P100, I've got the following numbers for your snippet:
```
In [23]: %timeit loop(lambda: dst_fp16.copy_(src_fp16))
312 ms ± 633 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [24]: %timeit loop(lambda: dst_fp32.copy_(src_fp32))
313 ms ± 1.11 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
which seem to indicate that, for at least a P100 GPU using CUDA 10, there seems to be no difference in time for fp16 and fp32 copy

**Comment by ThisIsIsaac** [comment]
@fmassa I am testing on V100. 

The number of elements of fp16 and fp32 array are the same, which means fp16 array is half the size of fp32. Fp16 broadcasts to tensor of size 2GB (since half is 2 bytes) while fp32 to size of 4GB (since float is 4 bytes). It is interesting that you got different results from me, but it seems to indicate that on P100, fp16 is half as fast as fp32.

**Comment by fmassa** [comment]
@ThisIsIsaac yeah, I just looked a bit more carefully at your snippet, and I saw that the operations that are being performed are on tensors of the same size and different dtypes, so yeah, having the same speed is not expected.

**Comment by ThisIsIsaac** [comment]
I looked at the kernels and the problem seems to be wasted global memory bandwidth. Currently, each thread of a warp only reads one element regardless of the type. This is fine ( although potentially not optimal ) when the type is float, since size of float is 4 bytes. Global read and write granularity is 128 bytes and it is broadcasted to all thread in a warp (there are 32 threads in the warp ). Since each thread is reading / writing consecutive float element, we are taking advantage of full 128 bytes (4 bytes per element * 1 element per thread * 32 threads = 128 bytes,  which is equal to global read/write granularity ). However, when we are using fp16, then the warp is only using half the bandwidth ( 2 bytes per element * 1 element per thread * 32 threads = 64 bytes ). 

So by conditioning on the type and having each thread process 2 items at a time instead of 1 ( so now we have: 2 bytes per element * 2 element per thread * 32 threads = 128 bytes ) we can use the full 128 byes of read and write, doubling the speed of any element-wise operations for fp16 (and types with size smaller than 4 such as int8)

**Comment by fmassa** [comment]
cc @VitalyFedyunin  @ngimel what are your thoughts on this?

**Comment by ngimel** [comment]
I'm skeptical of the "doubling" claim - similar to numbers that @colesbury posted, we typically see half operations go at ~80% of fp32 rate, due to caching. 
Replacing single element accesses with multiple element accesses (e.g. __half2 reads/writes) as suggested here is going to bring perf benefits (not doubling, but still sizable), however, it will be at the cost of a considerable code complexity - e.g. this optimization can be used only if `OffsetCalculator.get` produces contiguous offsets for linead_idx and linear_idx+1 argument values https://github.com/pytorch/pytorch/blob/cfd8c58b453c42bedaeaae50411637d85feee904/aten/src/ATen/cuda/detail/OffsetCalculator.cuh#L37, and also that offset for `linear_idx` is properly aligned which may or may not be true depending on tensor data_ptr, dimensions and strides. I for one would not want to review that PR ;-). Any perf benefits should be evaluated against this code complexity and potential for hard to find bugs (also, increased binary size since there would be special kernels to handle "fast" paths, and current fallback kernels), and like @colesbury my initial reaction is "no".

**Comment by ngimel** [comment]
> @fmassa Thanks for the clarification. The problem is Waveglow requires Pytorch apex plugin, but it doesn't work with the newest pytorch ( only works with `1.0.0`).
@ThisIsIsaac if you notice apex not working with the newest pytorch, please file apex issues. It is regularly tested against latest pytorch, but of course some bugs could go unnoticed.

**Comment by ThisIsIsaac** [comment]
@ngimel Thanks for your feedback.

> I'm skeptical of the "doubling" claim - similar to numbers that @colesbury posted, we typically see half operations go at ~80% of fp32 rate, due to caching.

I also observed fp16 display ~75% throughput of fp32, so I expect about 25% increase in throughput on my V100. The doubling claim was a reflection on @fmassa 's observation his P100. 

>  it will be at the cost of a considerable code complexity

I think it would be of huge benefit to at least give it a shot for kernels that are bottlenecks. We observed that Waveglow spends about 20~30% of time in element-wise operations (it spends more time in element-wise operations than in any cuDNN or cuBLAS kernels). This may be observed from other algorithms since element-wise operations and copies are the most common types of operations. **However, if you deem the throughput increase unworthy of the code complexity, I completely understand as you are the one who would have to maintain the repo in the future.** 

---

On a different note, I only see `half2` being used in [fastSpecializedAtomicAddKernel](https://github.com/pytorch/pytorch/blob/056a033cdc38a7e16d3a59a1c7d43daf2758d064/aten/src/ATen/native/cuda/KernelUtils.cuh). Is the reason we don't use `half2` also to avoid code complexity? Or are there plans to utilize `half2` in the future?

**Comment by ngimel** [comment]
I've looked a bit more, and `elementwise_kernel` already has a `vt` parameter controlling how many elements each thread processes, so speeding up half operations might be as easy as changing this parameter https://github.com/pytorch/pytorch/blob/1aae4b02df797f98239fde511575c96b8d6806da/aten/src/ATen/native/cuda/Loops.cuh#L67-L80. @ThisIsIsaac is it something you want to experiment with? 
As for `half2`, this is subject to all alignment and contiguity complications I mentioned before :-( The reason we had to use half2 in atomicAdd is that regular atomic add for halfs is very (order of magnitude) slow, so e.g. upsampling backward kernels using atomics were 3-5 times slower than corresponding kernels for float, leading to large end-to-end slowdowns. If you look at it, fastSpecializedAtomicAdd has 4 arguments, all of them necessary, instead of just 2 for a regular atomicAdd, so there's code complexity right there.
